Mathlib/MeasureTheory/MeasurableSpace/CountablyGenerated.lean

```diff
@@ -99,6 +117,140 @@ instance [MeasurableSpace α] [CountablyGenerated α] [MeasurableSpace β] [Coun
+    (h : (∀ s, MeasurableSet s → (x ∈ s → y ∈ s))) : x = y := hs.separates _ _ h
+
+theorem exists_measurableSet_of_neq [MeasurableSpace α] [SeparatesPoints α] {x y : α}
+    (h : x ≠ y) : ∃ s, MeasurableSet s ∧ x ∈ s ∧ y ∉ s := by
+  contrapose! h
+  exact separatesPoints_def h
+
+/-- If the measurable space generated by `S` separates points,
+then this is witnessed by sets in `S`. -/
+theorem separating_of_generateFrom (S : Set (Set α))
+    [h : @SeparatesPoints α (generateFrom S)] :
+    ∀ x y : α, (∀ s ∈ S, x ∈ s ↔ y ∈ s) → x = y := by
+  letI := generateFrom S
+  intros x y hxy
+  rw [← forall_generateFrom_mem_iff_mem_iff] at hxy
+  exact separatesPoints_def $ fun _ hs => (hxy _ hs).mp
+
+theorem SeparatesPoints.mono {m m' : MeasurableSpace α} [hsep : @SeparatesPoints _ m] (h : m ≤ m') :
+    @SeparatesPoints _ m' := @SeparatesPoints.mk _ m' fun _ _ hxy ↦
+    @SeparatesPoints.separates _ m hsep _ _ fun _ hs ↦ hxy _ (h _ hs)
+
+instance (priority := 100) Subtype.separatesPoints [MeasurableSpace α] [h : SeparatesPoints α]
+    {s : Set α} : SeparatesPoints s :=
+  ⟨fun _ _ hxy ↦ Subtype.val_injective $ h.1 _ _ fun _ ht ↦ hxy _ $ measurable_subtype_coe ht⟩
+
+instance (priority := 100) separatesPoints_of_measurableSingletonClass [MeasurableSpace α]
+    [MeasurableSingletonClass α] : SeparatesPoints α := by
+  refine' ⟨fun x y h ↦ _⟩
+  specialize h _ (MeasurableSet.singleton x)
+  simp_rw [mem_singleton_iff, forall_true_left] at h
+  exact h.symm
+
+instance(priority := 100) [MeasurableSpace α] {s : Set α}
+    [h : CountablyGenerated s] [SeparatesPoints s] :
+    HasCountableSeparatingOn α MeasurableSet s := by
+  suffices HasCountableSeparatingOn s MeasurableSet univ from this.of_subtype fun _ ↦ id
+  rcases h.1 with ⟨b, hbc, hb⟩
+  refine ⟨⟨b, hbc, fun t ht ↦ hb.symm ▸ .basic t ht, ?_⟩⟩
+  rw [hb] at ‹SeparatesPoints s›
+  convert separating_of_generateFrom b
+  simp
+
+variable (α)
+
+/-- If a measurable space admits a countable separating family of measurable sets,
+there is a countably generated coarser space which still separates points. -/
+theorem exists_countablyGenerated_le_of_hasCountableSeparatingOn [m : MeasurableSpace α]
+    [h : HasCountableSeparatingOn α MeasurableSet univ] :
+    ∃ m' : MeasurableSpace α, @CountablyGenerated _ m' ∧ @SeparatesPoints _ m' ∧ m' ≤ m := by
+  rcases h.1 with ⟨b, bct, hbm, hb⟩
+  refine' ⟨generateFrom b, _, _, generateFrom_le hbm⟩
+  · use b
+  refine' @SeparatesPoints.mk _ (generateFrom b) fun x y hxy ↦ hb _ trivial _ trivial _
+  intro s hs
+  use hxy _ (measurableSet_generateFrom hs)
+  contrapose
+  exact hxy _ (measurableSet_generateFrom hs).compl
+
+open scoped Classical
+
+open Function
+
+/-- A map from a measurable space to the Cantor space `ℕ → Bool` induced by a countable
+sequence of sets generating the measurable space. -/
+noncomputable
+def map_nat_bool_of_countablyGenerated [MeasurableSpace α] [CountablyGenerated α] (x : α) (n : ℕ) :
```

Comment:
Perhaps name it `map_nat_bool` instead to shorten the name, omitting `of_contablyGenerated`? (only if you don't expect a similar map to be useful under different hypotheses).

Reply:
Actually I just realized that since it is a def, the name I suggest should not be `map_nat_bool` but `mapNatBool`.